src/Features/CSharp/Portable/Diagnostics/Analyzers/TypeSyntaxSimplifierWalker_Scopes.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.SimplifyTypeNames
+{
+    internal partial class TypeSyntaxSimplifierWalker : CSharpSyntaxWalker
+    {
+        private readonly CSharpSimplifyTypeNamesDiagnosticAnalyzer _analyzer;
+        private readonly SemanticModel _semanticModel;
+        private readonly OptionSet _optionSet;
+        private readonly bool _preferPredefinedTypeInDecl;
+        private readonly bool _preferPredefinedTypeInMemberAccess;
+        private readonly CancellationToken _cancellationToken;
+
+        /// <summary>
+        /// Set of all the type names that this compilation knows about (both through source and
+        /// metadata).  This is used so that when we see <c>Y.Z</c> we can know if <c>Y</c> could be
+        /// a type and thus if we should try to simplify that to <c>X.Z</c> (if <c>X</c> is a base
+        /// class of <c>Y</c>).
+        /// </summary>
+        private readonly HashSet<string> _compilationTypeNames;
+
+        /// <summary>
+        /// Set of type and namespace names that have an alias associated with them.  i.e. if the
+        /// user has <c>using X = System.DateTime</c>, then <c>DateTime</c> will be in this set.
+        /// This is used so we can easily tell if we should try to simplify some identifier to an
+        /// alias when we encounter it.
+        /// </summary>
+        private HashSet<string> _aliasedSymbolNames;
+
+        /// <summary>
+        /// Set of types and namespace names currently in scope based on the usings/namespaces we're
+        /// inside of.  This can be used to tell if we can simplify <c>X.Y</c> (in a types-only
+        /// context) to just<c>Y</c>.  If there is no declaration in scope called <c>Y</c> we don't
+        /// have to bother checking.
+        /// </summary>
+        private HashSet<string> _declarationNamesInScope;
```

Comment:
> If there is no declaration in scope called `Y; we don't have to bother checking.

How much did we gain from this specifically? If the benefit is substantial, do you know why #19258 didn't already fix it? The same applies for `_staticNamesInScope`.

Reply:
Same issue here. The previous prs help, but still end up computing a *ton* of redundant values. 

The idea of this pr is to do the computation only as we enter compilations/namespaces/types.  That happens vastly less than the number of identifiers we encounter that we attempt to simplify.